network/lwip2transport/device.go

```diff
@@ -0,0 +1,211 @@
+	}
+}
+
+// MTU implements [network.IPDevice]. It returns the maximum buffer size of a single IP packet that can be processed by
+// this device.
+func (d *lwIPDevice) MTU() int {
+	return packetMTU
+}
+
+// forwardOutgoingIPPacket writes an IP packet response `b` to this device. The packet can be read by calling the Read
+// function, or it can be redirected to an [io.Writer] if the WriteTo function has been called. forwardOutgoingIPPacket
+// blocks until the packet is successfully consumed by a Read or WriteTo.
+//
+// forwardOutgoingIPPacket can be used as an output function for lwIP.
+//
+// forwardOutgoingIPPacket might be called by multiple goroutines (for example, when multiple UDP packets arrive at the
+// same time). We sequentialize the calls by using channels, if performance issues arise in the future, we can use
+// other more performant but more error-prone methods (e.g. the [sync] package) to resolve them.
+func (d *lwIPDevice) forwardOutgoingIPPacket(b []byte) (int, error) {
+	if len(b) == 0 {
+		return 0, nil
+	}
+	select {
+	case d.rdBuf <- b:
+		select {
+		case n := <-d.rdN:
+			return n, nil
+		case <-d.done:
+			return 0, network.ErrClosed
+		}
+	case <-d.done:
+		return 0, network.ErrClosed
+	}
+}
+
+// Read implements [io.Reader] and [network.IPDevice]. It reads one IP packet from the TCP/UDP response, blocking until
+// a packet arrives or this device is closed. If a packet is too long to fit in the supplied buffer `p`, the excess
+// bytes are discarded.
+//
+// Read returns [io.EOF] error if this device is closed or no more data is available.
+func (d *lwIPDevice) Read(p []byte) (int, error) {
+	select {
+	case s := <-d.rdBuf:
+		n := copy(p, s)
+		d.rdN <- n
+		return n, nil
+	case <-d.done:
+		return 0, io.EOF
+	}
+}
+
+// WriteTo implements [io.WriterTo]. It writes all IP packets from TCP/UDP responses to `w` until all data is written
+// or an error occurs. This function will not allocate any intermediate buffers.
+//
+// WriteTo returns the total number of bytes written and any error encountered during the write. If there are no more
+// data available, WriteTo returns nil error instead of [io.EOF].
+func (d *lwIPDevice) WriteTo(w io.Writer) (int64, error) {
+	nw := int64(0)
+	for {
+		select {
+		case s := <-d.rdBuf:
+			n, err := w.Write(s)
+			nw += int64(n)
+			select {
+			case d.rdN <- n:
+				if err != nil {
+					return nw, err
+				}
+			case <-d.done:
+				return nw, nil
+			}
+		case <-d.done:
+			return nw, nil
+		}
+	}
+}
+
+// Write implements [io.Writer] and [network.IPDevice]. It writes a single IP packet to this device. The device will
+// then translate the IP packet into a TCP or UDP traffic.
+//
+// Write returns [network.ErrClosed] if this device is already closed.
+// Write returns [network.ErrMsgSize] if len(b) > [MTU].
+func (d *lwIPDevice) Write(b []byte) (int, error) {
+	select {
+	case <-d.done:
+		return 0, network.ErrClosed
+	default:
+	}
+	if len(b) > d.MTU() {
```

Comment:
Do we need to force this failure? What happens to the Write below if the buffer is too large?

No need to block this PR, but that's something to check when we add tests. It may be the case that the stack does IP fragmentation, in which case this may be a regression.

The alternative would be to not add this check, and perhaps check the Write error.

Reply:
Good point. Removed it for now. Let's see how lwip handles it in the test.